app/src/main/java/org/mozilla/reference/browser/tabs/SyncedTabsFragment.kt

```diff
@@ -0,0 +1,271 @@
+import mozilla.components.concept.sync.AccountObserver
+import mozilla.components.concept.sync.AuthType
+import mozilla.components.concept.sync.Device
+import mozilla.components.concept.sync.OAuthAccount
+import mozilla.components.feature.syncedtabs.SyncedTabsFeature
+import mozilla.components.service.fxa.SyncEngine
+import mozilla.components.service.fxa.manager.FxaAccountManager
+import mozilla.components.service.fxa.manager.SyncEnginesStorage
+import mozilla.components.service.fxa.sync.SyncReason
+import mozilla.components.service.fxa.sync.SyncStatusObserver
+import org.mozilla.reference.browser.R
+import org.mozilla.reference.browser.ext.components
+
+// The tricky part in this class is to handle all possible Sync+FxA states:
+//
+// - No Sync account
+// - Connected to FxA but not Sync (impossible state on mobile at the moment).
+// - Connected to Sync, but needs reconnection.
+// - Connected to Sync, but tabs syncing disabled.
+// - Connected to Sync, but tabs haven't been synced yet (they stay in memory after the first sync).
+// - Connected to Sync, but only one device in the account (us), so no other tab to show.
+// - Connected to Sync.
+
+class SyncedTabsFragment : Fragment(), View.OnClickListener {
+    private lateinit var accountManager: FxaAccountManager
+    private lateinit var syncedTabsFeature: SyncedTabsFeature
+    private lateinit var recyclerView: RecyclerView
+    private val scope = CoroutineScope(Dispatchers.IO)
+
+    override fun onCreateView(
+        inflater: LayoutInflater,
+        container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View {
+        return inflater.inflate(R.layout.fragment_synced_tabs, container, false)
+    }
+
+    override fun onViewCreated(rootView: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(rootView, savedInstanceState)
+
+        accountManager = requireContext().components.backgroundServices.accountManager
+        syncedTabsFeature = requireContext().components.backgroundServices.syncedTabs
+
+        recyclerView = rootView.findViewById(R.id.synced_tabs_list)
+        recyclerView.layoutManager = LinearLayoutManager(requireContext())
+
+        // Listen to Sync/FxA events so we can update the UI accordingly.
+        accountManager.registerForSyncEvents(
+            SyncedTabsSyncObserver(this),
+            owner = this,
+            autoPause = true
+        )
+        accountManager.register(SyncedTabsAccountObserver(this))
+        updateUI()
+    }
+
+    override fun onResume() {
+        super.onResume()
+        // Kick off a sync in the background so we get the freshest tabs.
+        scope.launch {
+            accountManager.syncNowAsync(SyncReason.User).await()
+        }
+    }
+
+    fun updateUI() {
+        val accountManager = requireContext().components.backgroundServices.accountManager
+        val account = accountManager.authenticatedAccount()
+
+        if (account != null && !accountManager.accountNeedsReauth()) {
```

Comment:
We can make these if-else blocks a bit more readable with a single `when` block.